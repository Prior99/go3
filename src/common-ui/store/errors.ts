import { observable, computed, action } from "mobx";
import { bindAll } from "lodash-decorators";
import { component, inject, initialize } from "tsdi";

interface ApiError {
    message: string;
}

@bindAll()
@component
export class ErrorStore {
    @observable public errors: ApiError[] = [];

    @action public report(error: ApiError) {
        if (this.errors.find(other => other.message === error.message)) {
            return;
        }
        this.errors.push(error);
    }

    @action public dismiss() {
        this.errors.pop();
    }

    @computed
    public get latestError() {
        return this.errors[this.errors.length - 1];
    }
}
