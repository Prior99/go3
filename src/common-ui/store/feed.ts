import { observable, computed, action } from "mobx";
import { bind } from "bind-decorator";
import { component, initialize, inject } from "tsdi";

import { FeedItem, Feed } from "../../common";

@component
export class FeedStore {
    @inject private feed: Feed;

    @observable private feedItems: FeedItem[] = [];

    @initialize @bind @action
    public async loadFeed() {
        this.feedItems = await this.feed.getFeed();
    }

    @computed public get all() {
        return this.feedItems;
    }
}
