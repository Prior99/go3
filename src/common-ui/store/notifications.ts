import { observable, action, autorun } from "mobx";
import { component, inject, initialize } from "tsdi";
import { bind } from "decko";

import { Game, formatPosition } from "../../common";
import { drawBoard, ServiceWorkerManager } from "..";
import { GamesStore, OwnUserStore } from ".";

@component({ name: "NotificationsStore", eager: true })
export class NotificationsStore {
    @inject private games: GamesStore;
    @inject private ownUser: OwnUserStore;
    @inject("ServiceWorkerManager") private serviceWorkerManager: ServiceWorkerManager;

    @observable private forbidden = false;
    @observable private lastTurns: Map<string, number> = new Map();
    private useServiceWorker = false;

    public useServiceWorkerApi() {
        this.useServiceWorker = true;
    }

    @initialize
    private async init() {
        const permission = await Notification.requestPermission();
        if (permission !== "granted") {
            this.forbidden = true;
        }
        autorun(() => {
            if (this.useServiceWorker) {
                return;
            }
            this.checkNotifications();
        });
    }

    @bind private notify(game: Game) {
        const opponent = game.getOpponent(this.ownUser.user.id).name;
        const { turn } = game;
        const position = formatPosition(game.currentBoard.toPos(game.currentBoard.placedAt), game.boardSize);
        const canvas = document.createElement("canvas");
        canvas.width = 256;
        canvas.height = 256;
        drawBoard(canvas, game.currentBoard);
        const title = "It's your turn";
        const options = {
            body: `${opponent} played turn ${turn} at ${position} and now it's your turn.`,
            icon: canvas.toDataURL(),
        };
        if (this.useServiceWorkerApi) {
            this.serviceWorkerManager.registration.showNotification(title, {
                ...options,
                badge: "/badge.png",
                vibrate: [100, 100, 250],
                sound: "/notify.mp3",
                timestamp: game.currentBoard.created.getTime(),
                tag: game.id,
                data: game.id,
            } as any);
        } else if (!this.forbidden) {
            new Notification(title, options); // tslint:disable-line
        }
    }

    @bind public checkNotifications() {
        this.games.all.forEach(game => {
            const shouldNotify = this.lastTurns.has(game.id) &&
                game.turn > this.lastTurns.get(game.id) &&
                game.currentUser.id === this.ownUser.user.id;
            if (shouldNotify) {
                this.notify(game);
            }
            this.lastTurns.set(game.id, game.turn);
        });
    }
}
