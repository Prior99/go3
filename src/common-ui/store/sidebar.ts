import { observable, computed, action } from "mobx";
import { bindAll } from "lodash-decorators";
import { component, initialize } from "tsdi";

import { breakpointL } from "../breakpoints";

@component
@bindAll()
export class SidebarStore {
    @observable public visibilityToggled = false;
    @observable public alwaysOpen = this.calculateAlwaysOpen();

    private calculateAlwaysOpen() {
        return window.innerWidth >= breakpointL;
    }

    private onWindowResize() {
        this.alwaysOpen = this.calculateAlwaysOpen();
    }

    @initialize public init() {
        window.addEventListener("resize", this.onWindowResize);
    }

    @computed public get visible() {
        return this.visibilityToggled || this.alwaysOpen;
    }

    @action public toggleVisibility() {
        this.visibilityToggled = ! this.visibilityToggled;
    }
}
