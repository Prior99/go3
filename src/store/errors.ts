import { observable, computed, action } from "mobx";
import { bind } from "bind-decorator";
import { component, inject, initialize } from "tsdi";

interface ApiError {
    message: string;
}

@component
export class ErrorStore {
    @observable public errors: ApiError[] = [];

    @bind @action
    public dismiss() {
        this.errors.pop();
    }

    @computed
    public get latestError() {
        return this.errors[this.errors.length - 1];
    }
}
