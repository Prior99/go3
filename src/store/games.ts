import { observable, computed, action, extendObservable } from "mobx";
import { bind } from "bind-decorator";
import { LoginStore } from ".";
import { History } from "history";
import { component, inject, initialize } from "tsdi";
import { Game } from "../models";
import { Games, Users } from "controllers";
import { Color } from "../board-color";
import { UsersStore } from "./users";
import { routeGame } from "../routing/index";
import * as pathToRegexp from "path-to-regexp";

function parseGameId(url: string) {
    const result = pathToRegexp(routeGame.pattern).exec(url);
    if (!result) {
        return;
    }
    return result[1];
}

@component("GamesStore")
export class GamesStore {
    @inject private browserHistory: History;
    @inject private gamesController: Games;
    @inject private usersController: Users;
    @inject("LoginStore") private login: LoginStore;
    @inject private users: UsersStore;

    @observable private games: Map<string, Game> = new Map();
    @observable public loading = false;
    @observable public currentGameId: string;

    private refreshInterval: any;

    @initialize
    private async initialize() {
        if (this.login.loggedIn) {
            await this.loadGames();
        }
        this.browserHistory.listen(this.refreshGameId);
        this.refreshGameId();
    }

    @bind
    private async refreshGameId() {
        this.currentGameId = parseGameId(this.browserHistory.location.pathname);
        if (this.currentGameId) {
            this.loadBoards(this.currentGameId);
            this.refreshInterval = setInterval(this.refreshBoards, 1000);
        }
    }

    @bind
    private async refreshBoards() {
        if (this.currentGameId) {
            const newBoards = await this.gamesController.listBoards(this.currentGameId, this.currentGame.turn);
            this.games.get(this.currentGameId).boards.push(...newBoards);
            if (this.currentGame.consecutivePasses >= 2) {
                await this.loadGame(this.currentGameId);
                await this.loadBoards(this.currentGameId);
            }
            if (this.currentGame.over) {
                clearInterval(this.refreshInterval);
            }
            return;
        }
        clearInterval(this.refreshInterval);
    }

    @bind
    private async storeGame(game: Game) {
        this.games.set(game.id, game);
        await Promise.all(game.participants.map(({ user }) => this.users.load(user.id)));
    }

    @bind @action
    public async loadGames() {
        this.loading = true;
        if (this.login.loggedIn) {
            const games = await this.usersController.listGames(this.login.userId);
            games.forEach(this.storeGame);
        }
        this.loading = false;
    }

    @bind @action
    public async loadGame(id: string) {
        this.games.set(id, await this.gamesController.getGame(id));
    }

    @bind @action
    public async createGame(userId: string, size: number) {
        this.loading = true;
        const game = await this.gamesController.createGame({
            participants: [
                {
                    user: { id: userId },
                    color: Color.BLACK,
                },
                {
                    user: { id: this.login.userId },
                    color: Color.WHITE,
                },
            ],
            boardSize: size,
        } as Game);
        this.storeGame(game);
        this.loading = false;
        return game;
    }

    @bind @action
    private async loadBoards(gameId: string) {
        const boards = await this.gamesController.listBoards(gameId);
        this.games.get(gameId).boards = boards;
    }

    @computed
    public get all(): Game[] {
        return Array.from(this.games.values());
    }

    @bind
    public byId(id: string) {
        return this.games.get(id);
    }

    @bind @action
    public async turn(game: Game, index: number) {
        const board = await this.gamesController.turn(game.id, index);
        game.boards.push(board);
    }

    @bind @action
    public async pass(game: Game) {
        const board = await this.gamesController.pass(game.id);
        game.boards.push(board);
        if (game.consecutivePasses >= 1) {
            await this.loadGame(game.id);
            await this.loadBoards(game.id);
        }
    }

    @computed
    public get currentGame() {
        if (!this.currentGameId) {
            return;
        }
        return this.byId(this.currentGameId);
    }

    @computed
    public get ownColor() {
        if (this.login.userId === this.currentGame.whiteUser.id) {
            return Color.WHITE;
        }
        if (this.login.userId === this.currentGame.blackUser.id) {
            return Color.BLACK;
        }
        return;
    }
}
