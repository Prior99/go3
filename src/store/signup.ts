import { observable, computed, action } from "mobx";
import { bind } from "bind-decorator";
import { LoginStore, OwnUserStore } from ".";
import { History } from "history";
import { component, inject } from "tsdi";
import { Users } from "controllers";

@component
export class SignupStore {
    @inject private login: LoginStore;
    @inject private users: Users;
    @inject private ownUser: OwnUserStore;

    @observable public signupResult: Boolean;

    @bind @action
    public async signup(email: string, password: string, name: string) {
        const body = { email, password };
        const response = await this.users.createUser({ email, password, name });
        if (response) {
            await this.login.login(email, password);
            await this.ownUser.loadUser();
        }
        return response;
    }
}
