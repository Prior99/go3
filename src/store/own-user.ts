import { observable, computed, action } from "mobx";
import { bind } from "bind-decorator";
import { component, initialize, inject } from "tsdi";
import { User } from "models";
import { Users } from "controllers";
import { LoginStore } from "store";

@component
export class OwnUserStore {
    @inject private users: Users;
    @inject private login: LoginStore;

    @observable public user: User;

    @initialize @bind @action
    public async loadUser() {
        const response = await this.users.getUser(this.login.userId);
        if (response) {
            this.user = response as User;
        }
    }
}
