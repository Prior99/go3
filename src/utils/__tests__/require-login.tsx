import * as React from "react";
import { requireLogin } from "..";
import { createRenderer, ShallowRenderer } from "react-test-renderer/shallow";
import { LoginStore } from "store";
import { Provider } from "mobx-react";

class SomeComponent extends React.PureComponent<{}> {
    public render() {
        return (
            <div>Some Component</div>
        );
    }
}

let DecoratedComponent: React.ComponentClass<{}>;
let renderer: ShallowRenderer;

beforeEach(() => {
    renderer = createRenderer();
    DecoratedComponent = requireLogin(SomeComponent);
});

test("`requireLogin()` redirects to the `login()` route if not logged in", () => {
    const current = renderer.render(<DecoratedComponent />);
    expect(current).toMatchSnapshot();
});

test("`requireLogin()` does not redirect if logged in", () => {
    const apiStore = tsdi.get(LoginStore);
    apiStore.authToken = "some-auth-token";
    const current = renderer.render(<DecoratedComponent />);
    expect(current).toMatchSnapshot();
});
