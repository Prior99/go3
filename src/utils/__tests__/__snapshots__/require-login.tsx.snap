// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`\`requireLogin()\` does not redirect if logged in 1`] = `<SomeComponent />`;

exports[`\`requireLogin()\` redirects to the \`login()\` route if not logged in 1`] = `
<Redirect
  push={false}
  to="/login"
/>
`;
